### DISCUSSION TITLE
Trouble connecting to indigo service using ketcher-react

### DISCUSSION THREAD

**Discussion by bhans-uvic** [epam/ketcher#5515] [Q&A]
I am using ketcher-react for my project and want to connect to indigo-service to use the calculate functionality. I am using NextJS as framework of preference but I am pretty sure it should work with other frameworks as well. Does anyone have any ideas on how to make it work or how can I use the ketcher object to fetch information from the editor?

I have the indigo service running as a docker container on my local machine on port 8002.

Here's my component code:

```
const SketcherEditor = () => {
  //const structServiceProvider = new StandaloneStructServiceProvider();
  const structServiceProvider = new RemoteStructServiceProvider(
    process.env.REACT_APP_API_PATH!,
    {
      "custom header": "value", // optionally you can add custom headers object
    },
  );

  return (
    <Editor
      errorHandler={(message: string) => console.log(message)}
      staticResourcesUrl={process.env.PUBLIC_URL!}
      structServiceProvider={structServiceProvider}
      onInit={(ketcher: Ketcher) => {
        window.ketcher = ketcher;
        window.parent.postMessage(
          {
            eventType: "init",
          },
          "*",
        );
      }}
    />
  );
};
```

The evironment variables are:
REACT_APP_API_PATH="/v2/indigo"
PUBLIC_URL="http://localhost:8002"

**Comment by rrodionov91** [comment]
Hi @bhans-uvic 
Thank you for raising the issue.
Look's like your code is correct. I did the same and it works:
<img width="1041" alt="image" src="https://github.com/user-attachments/assets/dc03386b-2525-42e8-a0f7-1f98d18fb2de">

We have an example of ketcher usage with ssr, but for standalone indigo here https://github.com/epam/ketcher/blob/master/example-ssr/src/app/editor.tsx

Could you please explain the problem in more details?

**Reply by bhans-uvic** [reply]
Hi @rrodionov91 ,
Thank you for replying. I realized I was missing the api_path in the URL.
I was sticking to the ketcher-react document and doing one of the following:

<Editor staticResourcesUrl={process.env.PUBLIC_URL} apiPath={link to Indigo service} />
    or
http://localhost:3000/?api_path={link to Indigo service}

Soon after posting my question I took a closer look at your demo on the [epam](https://lifescience.opensource.epam.com/KetcherDemo/index.html?api_path=/v2) website and realized that I have to do both, the 'or' is a misdirection.

I was also wondering if I can get the results of the /calculate request through the window.ketcher object or not? Is that a possibility or are the API requests abstracted away from the object itself?

**Reply by rrodionov91** [reply]
Hi @bhans-uvic 
Sorry for late response
As I can see /calculate is not a part of ketcher api currently and we have plans for it in near future. It would be great if you have ability to implement it and create PR to ketcher repository. We can navigate you in codebase.

**Reply by bhans-uvic** [reply]
I would be happy to contribute.
I do want to clarify something that's confusing to me though, wouldn't RemoteStructService.calculate() work for this? I haven't tested this but I am assuming passing the results of ketcher.getSmarts() (or something similar) and data: ['molecular-weight', 'most-abundant-mass', 'monoisotopic-mass', 'gross', 'mass-composition'] should work?

**Reply by rrodionov91** [reply]
You are looking into right place.

Ketcher uses RemoteStructService and StandaloneStructService to communicate with indigo(backend). However Ketcher does not use it directly because only instance of one of them can exist in Ketcher.
You can take a look packages/ketcher-core/src/application/ketcher.ts:427 (layout method) as a reference of how usually ketcher api implemented.

Please let me know in case of any questions.

**Reply by bhans-uvic** [reply]
That helped, thanks. Please do let me know if your team follows a specific contribution workflow/guide, otherwise I can create a PR with commits similar to the ones I am seeing in the log.

What are your thoughts on the following?
`async calculate(options?: CalculateData): Promise<CalculateResult> {
    if (window.isPolymerEditorTurnedOn) {
      throw new Error('Calculate is not available in macro mode');
    }
    return await this._indigo.calculate(this.#editor.struct(), options);
  }`

I ended up forking the project and trying it out, I am facing issues with tests when pushing though. Some tests doesn't run and the push stays in stasis for a long time, so I have created the PR by skipping the tests during push:
![Screenshot 2024-11-02 110703](https://github.com/user-attachments/assets/4908dfd3-44f7-4f7b-b082-aeac2246b503)
![Screenshot 2024-11-02 110715](https://github.com/user-attachments/assets/791d5a56-f939-4702-99ce-74db993e4efb)

**Reply by bhans-uvic** [reply]
Hi @rrodionov91, wanted to follow-up on my [PR](https://github.com/epam/ketcher/pull/5899). We have an event coming up where I would like to present our application. But, we need the calculate functionality for the ideal scenario. Would it be possible to have the API available in a minor release by next week?

**Reply by rrodionov91** [reply]
Hi @bhans-uvic
Thank you for PR. I just merged it. 
Next release date from master branch is on 1 December. I can create dev release for you.

**Reply by rrodionov91** [reply]
@bhans-uvic
We published 2.28.0-dev.1 with your changes. 
Please let me know if something else needed. And thank you again.

**Reply by bhans-uvic** [reply]
Thank you so much for the quick release @rrodionov91. Truly appreciate it.
You guys are just awesome for making this amazing codebase and application open source for everyone. Please continue the great work! 👍 💯
